presto-main/src/main/java/com/facebook/presto/sql/planner/PlanVariableAllocator.java

```diff
@@ -174,20 +132,16 @@ else if (expression instanceof GroupingOperation) {
         return newVariable(nameHint, type, suffix);
     }
 
-    public Symbol newSymbol(Field field)
-    {
-        String nameHint = field.getName().orElse("field");
-        return newSymbol(nameHint, field.getType());
-    }
-
     public VariableReferenceExpression newVariable(Field field)
     {
         return newVariable(field.getName().orElse("field"), field.getType(), null);
     }
 
     public TypeProvider getTypes()
     {
-        return TypeProvider.viewOf(symbols);
+        return TypeProvider.fromVariables(variables.entrySet().stream()
```

Comment:
I am a little worried about the performance of this change here. `getTypes` can be used quite frequently, each time we will make a copy of existing types. 

Can we change the `TypeProvider` to store only Map<String, Type> as well? It looks like we don't need to use Symbol there as well.  

The only other place copied the whole `TypeProvider` is `areExpressionsEquivalent` method. I am not sure if we need to convert the variable to input there though. I think variable name is unique already in `TypeProvider`, so I am not sure translate it to integer will help much in canonicalize the expression.

Reply:
I was hoping that JVM would be able to optimize this if it's actually going to be called way too many times. Totally didn't test it though.

Reply:
Added a commit to change `TypeProvider` to use `Map<String, Type>`.